66_HLSLBxDFTests/app_resources/test_components.hlsl

```diff
@@ -0,0 +1,378 @@
+struct TestNDF : TestBxDF<BxDF>
+{
+    using base_t = TestBxDFBase<BxDF>;
+    using this_t = TestNDF<BxDF, aniso>;
+    using traits_t = bxdf::traits<BxDF>;
+
+    virtual ErrorType compute() override
+    {
+        aniso_cache dummy;
+        iso_cache dummy_iso;
+
+        float32_t3 ux = base_t::rc.u + float32_t3(eps,0,0);
+        float32_t3 uy = base_t::rc.u + float32_t3(0,eps,0);
+
+        if NBL_CONSTEXPR_FUNC (traits_t::type == bxdf::BT_BRDF && traits_t::IsMicrofacet)
+        {
+            if NBL_CONSTEXPR_FUNC (aniso)
+            {
+                s = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy, cache);
+                sx = base_t::bxdf.generate(base_t::anisointer, ux.xy, dummy);
+                sy = base_t::bxdf.generate(base_t::anisointer, uy.xy, dummy);
+            }
+            else
+            {
+                s = base_t::bxdf.generate(base_t::isointer, base_t::rc.u.xy, isocache);
+                sx = base_t::bxdf.generate(base_t::isointer, ux.xy, dummy_iso);
+                sy = base_t::bxdf.generate(base_t::isointer, uy.xy, dummy_iso);
+            }
+        }
+        if NBL_CONSTEXPR_FUNC (traits_t::type == bxdf::BT_BSDF && traits_t::IsMicrofacet)
+        {
+            if NBL_CONSTEXPR_FUNC (aniso)
+            {
+                s = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u, cache);
+                sx = base_t::bxdf.generate(base_t::anisointer, ux, dummy);
+                sy = base_t::bxdf.generate(base_t::anisointer, uy, dummy);
+            }
+            else
+            {
+                s = base_t::bxdf.generate(base_t::isointer, base_t::rc.u, isocache);
+                sx = base_t::bxdf.generate(base_t::isointer, ux, dummy_iso);
+                sy = base_t::bxdf.generate(base_t::isointer, uy, dummy_iso);
+            }
+        }
+
+        // test jacobian is correct
+        // float32_t3x3 fromTangentSpace = base_t::anisointer.getFromTangentSpace();
+        // ray_dir_info_t tmpL;
+        // tmpL.setDirection(sampling::ProjectedSphere<float>::generate(base_t::rc.u));
+        // s = sample_t::createFromTangentSpace(tmpL, fromTangentSpace);
+        // tmpL.setDirection(sampling::ProjectedSphere<float>::generate(ux));
+        // sx = sample_t::createFromTangentSpace(tmpL, fromTangentSpace);
+        // tmpL.setDirection(sampling::ProjectedSphere<float>::generate(uy));
+        // sy = sample_t::createFromTangentSpace(tmpL, fromTangentSpace);
+
+        if (!(s.isValid() && sx.isValid() && sy.isValid()))
+            return BET_INVALID;
+
+        // TODO: add checks with need clamp trait
+        if (traits_t::type == bxdf::BT_BRDF)
+        {
+            if (s.getNdotL() <= bit_cast<float>(numeric_limits<float>::min))
+                return BET_INVALID;
+        }
+        else if (traits_t::type == bxdf::BT_BSDF)
+        {
+            if (hlsl::abs(s.getNdotL()) <= bit_cast<float>(numeric_limits<float>::min))
+                return BET_INVALID;
+        }
```

Comment:
VNDF generated H may produce L below the hemisphere even though V and H are correct.

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/236#discussion_r2644758827